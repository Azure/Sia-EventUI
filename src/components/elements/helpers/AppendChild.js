import React from 'react'

const AppendChild = (component, childToAppend) => React.cloneElement(
  component,
  {
    children: component.props.children
      ? Array.isArray(component.props.children)
        ? component.props.children.concat(childToAppend)
        : [component.props.children, childToAppend]
      : childToAppend
  }
)

export default AppendChild