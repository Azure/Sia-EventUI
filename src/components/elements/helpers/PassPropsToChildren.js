import React from 'react'

const PassPropsToChildren = (children, props) => React.Children.map(
  children,
  child => child.type && child.type !== 'div'
    ? React.cloneElement(child, props)
    : React.cloneElement(
      child,
      {
        children: PassPropsToChildren(
          child.props.children,
          props
        )
      }
    )
)

export default PassPropsToChildren
