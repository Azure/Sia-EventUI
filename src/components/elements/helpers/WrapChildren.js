import React from 'react'

export const WrapChildren = (component, wrapper) => React.cloneElement(
  component,
  {
    children: React.Children.map(
      component.props.children,
      (child, index) => wrapper(child, index)
    )
  }
)

export default WrapChildren
