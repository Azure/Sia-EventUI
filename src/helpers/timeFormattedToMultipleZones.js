import { DateTime } from 'luxon'
import _ from 'underscore'

const zones = [
  { shortname: 'PT', ianaZone: 'America/Los_Angeles'},  // PST https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
  { shortname: 'IST', ianaZone: 'Asia/Kolkata'},         // India's IANA zone https://en.wikipedia.org/wiki/Time_in_India
  { shortname: 'GMT', ianaZone: 'Etc/GMT'}
]

const format = Object.assign(DateTime.DATE_SHORT, DateTime.TIME_24_WITH_SECONDS)

export const timeFormattedToMultipleZones = (time, timezones = zones) => {
  let dateFormat = { year: "numeric", month: "numeric", day: "numeric" }
  let convertTimeToZone = (timepoint) => {
    timepoint.timeInZone = time.setZone(timepoint.ianaZone)
    return timepoint
  }

  let timeInMultipleZones = _.chain(timezones)
      .map(convertTimeToZone)
      .groupBy((timepoint) => timepoint.timeInZone.toLocaleString(dateFormat))
      .map((dateGroup) => {
        return _.map(dateGroup, (timepoint, index) => {
          debugger
          return index === 0 ?
                  timepoint.timeInZone.toLocaleString(format) + ' ' + timepoint.shortname :
                  timepoint.timeInZone.toLocaleString(DateTime.TIME_24_WITH_SECONDS) + ' ' + timepoint.shortname
        }).join(', ')
      })
      .compact()
      .value()
      .join('; ')
  // let timeInMultipleZones = _.chain(timezones).map(convertTimeToZone).value().join('; ')
debugger
  return timeInMultipleZones
}

export default timeFormattedToMultipleZones

// () => {
//   timeInZone.toLocaleString(format)
//
// return formatted_time + ' ' + timezone.shortname
// }

// dateGroup) => {
//   return _.map(dateGroup, (datetime, date) => {
//     debugger
//     return _.first(dateGroup[date]) === datetime ?
//               datetime.toLocaleString(format) :
//               datetime.toLocaleString(DateTime.TIME_24_WITH_SECONDS)
//   }).join(' lejoiner ')
// }
