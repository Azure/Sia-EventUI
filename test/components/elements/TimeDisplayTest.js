import React from 'react'
import { expect } from 'chai'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { DateTime } from 'luxon'


import { describeSnapshot } from 'test/helpers/describeSnapshot'

import {
  TimeDisplayComponent,
  parseTime,
  mapStateToProps
} from 'components/elements/TimeDisplay'

describe('TimeDisplay', function () {
  const validTime = DateTime.fromObject({
    year: 2018,
    month: 4,
    day: 25,
    hour: 15,
    minute: 39,
    second: 22,
    zone: 'America/Los_Angeles'
  })

  describeSnapshot(function () {
    it('should match snapshot', function () {
      expect(toJson(shallow(<TimeDisplayComponent time={validTime} timezones={['UTC']} />)))
        .to.matchSnapshot()
    })
  })

  describe('when time is null', function () {
    it('should return null', function () {
      expect(toJson(shallow(<TimeDisplayComponent time={null} timezones={['UTC']} />)))
        .to.matchSnapshot()
    })
  })

  describe('parseTime', function () {
    describe('when input is valid', function () {
      it('should return value when given a DateTime', function () {
        expect(parseTime(DateTime.utc(2018, 4, 25)).isValid).to.be.true
      })

      it('should return parsed ISO string when given ISO string', function () {
        const validIsoTime = '2008-09-15T15:53:00+05:00'

        expect(parseTime(validIsoTime).isValid).to.be.true
      })
    })

    describe('when input is invalid', function () {
      it('isValid should be false when given an integer', function () {
        expect(parseTime(123).isValid).to.be.false
      })

      it('isValid should be false when given a non-ISO time string', function () {
        expect(parseTime('not an ISO string').isValid).to.be.false
      })
    })
  })

  describe('mapStateToProps', function () {
    const validStateObject = {
      timePreferences: {
        displayTimezones: ['UTC', 'America/Los_Angeles', 'other', 'timezones']
      }
    }
    const validOwnProps = {
      time: 'time object'
    }

    it('should return the correct timezones', function () {
      const testState = mapStateToProps(
        validStateObject,
        validOwnProps
      )

      expect(testState.timezones).to.be.an('array')
        .to.have.same.members(validStateObject.timePreferences.displayTimezones)
    })

    it('should get time from ownProps', function () {
      const testState = mapStateToProps(
        validStateObject,
        validOwnProps
      )

      expect(testState.time).to.equal(validOwnProps.time)
    })
  })
})
