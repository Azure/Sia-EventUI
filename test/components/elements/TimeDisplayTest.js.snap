// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeDisplay snapshot should match snapshot 1`] = `
<div>
  <div
    key="UTC"
  >
    (
    UTC
    ): 
    04/25/2018, 22:39:22
  </div>
</div>
`;

exports[`TimeDisplay snapshot should match snapshot 2`] = `""`;
